### DISCUSSION TITLE
Custom loader - how to add files?

### DISCUSSION THREAD

**Discussion by rzvc** [vimeo/psalm#7974] [Plugin development]
I'm trying to add support for a custom loader, but what I have so far doesn't seem to work, and I can't find the documentation on this topic.

This is my code so far:
```php
namespace test\Psalm;

use PhpParser;
use Psalm\Plugin\EventHandler\AfterMethodCallAnalysisInterface;
use Psalm\Plugin\EventHandler\Event\AfterMethodCallAnalysisEvent;

class TestLoadFunction implements AfterMethodCallAnalysisInterface
{
	public static function afterMethodCallAnalysis(AfterMethodCallAnalysisEvent $event) : void
	{
		if ($event->getMethodId() == 'Foo::load' && isset($event->getExpr()->args[0]))
		{
			$request = $event->getExpr()->args[0]->value;

			if ($request instanceof PhpParser\Node\Scalar\String_)
			{
				$path = 'src/' . $request->value;

				// Tried this, but it didn't work:
				$codebase = $event->getCodebase();
				$codebase->addFilesToAnalyze([$path => $path]);
			}
		}
	}
}
```
Adding the file like that, results in an "Uncaught InvalidArgumentException: Could not get file storage for src/loaded.php", which gets displayed after the analysis is done.

Any ideas on how this can be achieved?

Thanks.

**Comment by orklah** [comment]
Can you describe what you're trying to do first? What is the error you're trying to solve?

**Reply by rzvc** [reply]
My library has a method called `load()`, which loads code. The problem is that I could `load('something')`, and whatever new functionality that 'something' should expose to the code, is not visible to Psalm.

I'm trying to solve this with a plugin, that figures out what load() would do and then mimics the behaviour of require_once, so new functionality would become available to the rest of the code, after that line.

**Reply by orklah** [reply]
It shouldn't be needed IMHO. Wherever you will use the class 'something', Psalm will try to autoload it. Either you use a composer autoloader and Psalm will use it as well, or you can provide a custom autoloader in your psalm's config and you don't need to use a plugin.

**Reply by rzvc** [reply]
I know, but this is mainly for utility functions, which can't be autoloaded. Currently I'm resorting to manually including them all from the entry point, but that won't warn me when the actual code is trying to use one of these functions without it being available.

**Reply by orklah** [reply]
I see. Can you try using Scanner::queueClassLikeForScanning instead?

**Reply by rzvc** [reply]
Just tried that, doesn't seem to work. I also tried to trigger `IncludeAnalyzer` manually, which seemed to do its thing (judging by the --debug-by-line output, but it still didn't have the same effect as doing it via `require_once()`.

**Reply by AndrolGenhald** [reply]
Don't know if this works for you, but I handle this by including files defining utility functions in composer.json under `["autoload"]["files"]`.

**Reply by rzvc** [reply]
I'm not using composer, but my framework can do that too, it's just that some of this stuff is only loaded when needed.

**Reply by orklah** [reply]
Scanner::queueClassLikeForScanning is the function used by the symfony plugin to load services so it should work somehow, but I don't know a lot about it

**Reply by rzvc** [reply]
Maybe it requires an autoloader to be active, so it works by tricking it. I will add one to my plugin test environment and see how that goes.

**Reply by rzvc** [reply]
Yup, that didn't work either. In fact I made it output everything that got requested and no classes were required (other than the Psalm ones). I'll have to look at their plugin and see what they're doing.
